### PR TITLE
Fix: Correct open Preview in FF using editor via CDN

### PR DESCRIPTION
**Pull request by KarolDawidziuk**
## What is the purpose of this pull request?

Bug fix

## Does your PR contain necessary tests?

All patches that change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [ ] Unit tests
- [x] Manual tests

## Did you follow the CKEditor 4 code style guide?

Your code should follow the guidelines from the [CKEditor 4 code style guide](https://github.com/ckeditor/ckeditor4/blob/major/dev/docs/codestyle.md) which helps keep the entire codebase consistent.

- [x] PR is consistent with the code style guide

## What is the proposed changelog entry for this pull request?

```
* [#4444](https://github.com/ckeditor/ckeditor4/issues/4444): Fixed: [Print](https://ckeditor.com/cke4/addon/print) / [Preview](https://ckeditor.com/cke4/addon/preview) Print and Preview are not opened correctly on FF due to cross-origin.
```

## What changes did you make?

FF opens the preview in a separate window to avoid an anchor problems. Unfortunately, this problem persists even without this fix. Using a section that uses `fixDomain` fixed the preview problem. Additionally, I've split this issue into another one regarding to problem with print.  

## Which issues does your PR resolve?

Closes [#4444](https://github.com/ckeditor/ckeditor4/issues/4444).

### CODE REVIEW

**Comment by Comandeer** [comment]
Rebased onto latest `master`.

**Review by Comandeer — CHANGES_REQUESTED** [review]
The fix works correctly, but it seems that the PR contains two separate fixes for different bugs (see my inline comments).

File: plugins/preview/plugin.js
Comment (KarolDawidziuk) [reply plugins/preview/plugin.js]: > Commenting out this function doesn't seem to change anything in a way in which printing works in Fx 🤔 However I'm testing on macOS, the behaviour can be different in other OSes.
I'm tested it on macOS and VBox Win 7 and still it's not working without fix.

>  If that's the case, please report the issue as it's connected with different plugin (print) and its handling of executing printing.
This problem with run print is related with change in [this](https://github.com/ckeditor/ckeditor4/pull/4758/files#diff-36cb2ca37dc3086bbe191b661d4f8923ff5872b765b8d537439a568da622580aR202) line. That's why I fix this in one PR.

**Comment by KarolDawidziuk** [comment]
Thank's @Comandeer for the review.
It's ready for another turn.

File: tests/plugins/preview/manual/previewcdn.html
Comment (Comandeer) [inline tests/plugins/preview/manual/previewcdn.html:15]: I'm not happy with this solution. It requires updating the commit hash on every change inside the PR.

Yet I don't see a better solution :thinking:

**Review by Comandeer — CHANGES_REQUESTED** [review]
I'm still for removing the whole part fixing the print in Firefox. It is a separate bug that should be reported as blocked on this one. Additionally it requires much more refactoring than it's currently done:

1. Current fix for print breaks boundaries between `preview` and `print` plugin. Probably `preview` plugin should expose some kind of public API to allow adding callback fired when preview window is loaded.
2. The fix breaks preview if the user first prints and then opens preview. In such a case previe option will also display print dialog. It's connected with the fact that the preview window is created only once and then reused every time users click a "Print Preview" or "Print" button.
3. The fix adds much complexity and seems messy. Replacing it with public API for adding callback should make the whole thing far more readable. Especially as we already have triggering print on preview `load` event, so we basically duplicating the logic.

So I'm for reporting the print issue and leaving only the preview fix here.

Additionally, manual tests kill IE8 for some reason.

**Comment by KarolDawidziuk** [comment]
Thanks for review @Comandeer.
After your advice, I've created a separated [issue](https://github.com/ckeditor/ckeditor4/issues/4790) regarding to problem with print on FF and this PR include fix only for open Preview.  

Beside of this I found another way to open editor from CDN without using iframe that which caused of failing test on IE(the browser was freezing). It's look like a little hacky way but for now it's only working solution on IE that I found.

File: tests/plugins/preview/manual/previewcdn.html
Comment (Comandeer) [inline tests/plugins/preview/manual/previewcdn.html:22]:
```suggestion
			head = document.head || document.getElementsByTagName( 'head' )[ 0 ];
```

However it would be more elegant to use our `CKEDITOR.dom` methods instead of native ones.

File: tests/plugins/preview/manual/previewcdn.html
Comment (Comandeer) [inline tests/plugins/preview/manual/previewcdn.html:14]: Are these steps necessary?

**Review by Comandeer — CHANGES_REQUESTED** [review]
Fix works good, I've added only some inline comments about manual test.

Additionally please fix issue with double commits.

File: tests/plugins/preview/manual/previewcdn.html
Comment (KarolDawidziuk) [reply tests/plugins/preview/manual/previewcdn.html]: Good point. Thanks 👍

File: tests/plugins/preview/manual/previewcdn.html
Comment (KarolDawidziuk) [reply tests/plugins/preview/manual/previewcdn.html:14]: Yes, because bender loads all scripts by default every time, so I delete all files and namespace so that the new instance doesn't accidentally use local files. Additionally, we are sure that the new editor instance is loaded from CDN.

**Comment by KarolDawidziuk** [comment]
Thanks @Comandeer for review. 
 It's ready for the next one.

**Comment by f1ames** [comment]
Rebased onto latest `master`.

File: tests/plugins/preview/manual/previewcdn.html
Comment (f1ames) [inline tests/plugins/preview/manual/previewcdn.html:44]: This is a bit problematic since we have basic, standard and full presets and manual tests run for such preset use default plugins. So for `full all` used during testing phase we would like to have all plugins which are default for this preset.

The solution could be reading `plugins` list from the initial editor instance (before destroying) and then pass it to new editor config. WDYT?

File: tests/plugins/preview/manual/previewcdn.html
Comment (f1ames) [inline tests/plugins/preview/manual/previewcdn.html:17]: This PR will be merged to `master` soon so maybe we could replace (just before merging) this URL to point to `master` branch?

Still this would mean we are not testing the current local editor version which might be confusing and is not wehat we usually want to achieve with manual tests :thinking:

Comment (f1ames) [reply tests/plugins/preview/manual/previewcdn.html:17]: I wonder couldn't we use `http://0.0.0.0:1030/apps/ckeditor/ckeditor.js` or `http://127.0.0.1:1030/apps/ckeditor/ckeditor.js` or `http://localhost:1030/apps/ckeditor/ckeditor.js` to simulate CORS (depending what is the current host used)?

From what I have checked (from `0.0.0.0`) it will give us CORS issue (so the same think we would like to simulate here I suppose):
image

And then we will have test using local version.

File: tests/plugins/preview/manual/previewcdn.md
Comment (f1ames) [inline tests/plugins/preview/manual/previewcdn.md:11]: > New browser tab with preview are opened correctly

This should be expected result after 2nd step - you click preview and new tab/window with correct content is shown.

> with no error in browser console

And this is expected result after 3rd step.

File: tests/plugins/preview/manual/previewcdn.md
Comment (f1ames) [inline tests/plugins/preview/manual/previewcdn.md:3]: This test and PR is about `preview` plugin so I;m not sure if `print` is needed :thinking:

File: tests/plugins/preview/manual/previewcdn.md
Comment (f1ames) [inline tests/plugins/preview/manual/previewcdn.md:16]: Again, why it's mentioning `print` if it's about `preview` plugin? :thinking:

**Review by f1ames — CHANGES_REQUESTED** [review]
The fix looks good. There are still some things to polish in manual test.

File: tests/plugins/preview/manual/previewcdn.html
Comment (KarolDawidziuk) [reply tests/plugins/preview/manual/previewcdn.html:44]: Good point. I didn't think about presets and your idea might work. Thanks!

File: tests/plugins/preview/manual/previewcdn.html
Comment (KarolDawidziuk) [reply tests/plugins/preview/manual/previewcdn.html:17]: It's look like your suggestion with simulate CORS locally works. I just need to change it a little and it should work. Thanks!

File: tests/plugins/preview/manual/previewcdn.md
Comment (KarolDawidziuk) [reply tests/plugins/preview/manual/previewcdn.md:11]: I'll correct it. Thanks.

File: tests/plugins/preview/manual/previewcdn.md
Comment (KarolDawidziuk) [reply tests/plugins/preview/manual/previewcdn.md:3]: My bad. I left it after the first PR version which included a problem with print and preview.

File: tests/plugins/preview/manual/previewcdn.md
Comment (KarolDawidziuk) [reply tests/plugins/preview/manual/previewcdn.md:16]: As I mentioned above i forgot remove it.

**Comment by KarolDawidziuk** [comment]
Thanks @f1ames for the review. 
It's ready for the next turn.

Changes:
* Changed places of expected results.
* Changed `print` to `preview` in manual test.
* Mixed CORS simulation.

> I wonder couldn't we use `http://0.0.0.0:1030/apps/ckeditor/ckeditor.js` or `http://127.0.0.1:1030/apps/ckeditor/ckeditor.js` or `http://localhost:1030/apps/ckeditor/ckeditor.js` to simulate CORS (depending what is the current host used)?
> 
> From what I have checked (from `0.0.0.0`) it will give us CORS issue (so the same think we would like to simulate here I suppose):
> image
> 
> And then we will have test using local version.


Unfortunately, this will not work when we are going to use a virtual machine because we'll only have `10.0.2.2:<port>` host. So after our F2F talk I added mixed solution. 
When a person runs the test in the virtual machine, editor it will be created from the master branch from the githack, otherwise CORS from the local editor will be simulated.

**Review by f1ames — CHANGES_REQUESTED** [review]
Manual test fails on IE8:
image

Apart from that it looks good. I did some small adjustments in manual test description and minor refactoring in test code itself.

**Comment by KarolDawidziuk** [comment]
Hi @f1ames  thanks for review. 
It's ready for another one.

**Comment by f1ames** [comment]
Rebased onto latest `master`.

**Pull request merged**
